File: frontend/src/lib/memorySuggestions.ts
/** Client-side memory/milestone detection (mirrors backend/memory_suggestions.py). */

import type { AppMemory } from './memoryTypes';

export type MemorySuggestionKind = 'milestone' | 'moment';

export interface MemorySuggestion {
  text: string;
  emoji: string;
  description?: string | null;
  ref?: string | null;
  kind?: MemorySuggestionKind;
  dateIso?: string;
  added?: boolean;
  dismissed?: boolean;
}

const SKIP = /^(\s*(τι|what|how|why|when|where|ποιος|ποια|ποιο|πώς|γιατί|πότε|που)\b|.*\?(.*\?)?$|(should i|τι να κάνω|help me|βοήθ|μπορείς|can you|tell me|πες μου|explain))/i;

const MILESTONE =
  /(πρώτ|πρωτ|1ο\s|1η\s|first\s+(time|steps?|word|tooth|smile|bath|haircut)|πρώτα\s+(βήματα|δοντάκ|χαμόγελ|λέξη|μπιμπίκο|κούρεμα)|πρωτα\s+(βηματα|δοντακ|χαμογελ|λεξη|μπικινι|κουρεμα))/i;

const MOMENT =
  /(σήμερα|σημερα|χθες|εχθές|εχες|today|yesterday).{0,50}(έκανε|εκανε|έκαν|εκαν|περπάτη|περπατη|walk|smil|κοιμ|slept|είπε|ειπε|said|έφαγε|εφαγε|ate|έπεσε|επεσε|fell|μπήκε|μπηκε|climb)/i;

const EMOTIONAL = /(🎉|💛|❤️|τόσο\s+όμορφ|so\s+(cute|proud|happy)|μπράβο|proud)/i;

const EMOJI_MAP: [RegExp, string][] = [
  [/(δοντ|tooth|teeth)/i, '🦷'],
  [/(χαμογ|smil)/i, '😊'],
  [/(βήμα|walk|step|περπα)/i, '🚶'],
  [/(μπάνι|bath)/i, '🛁'],
  [/(κοιμ|sleep)/i, '😴'],
  [/(μίλ|word|speak|said|είπε|babbl)/i, '👶'],
];

function normalize(s: string): string {
  return (s || '').trim().toLowerCase().replace(/\s+/g, ' ');
}

function emojiFor(text: string): string {
  for (const [pat, em] of EMOJI_MAP) {
    if (pat.test(text)) return em;
  }
  return '💛';
}

function isDuplicate(text: string, recent: Pick<AppMemory, 'text'>[]): boolean {
  const norm = normalize(text).slice(0, 100);
  if (!norm) return true;
  for (const m of recent) {
    const existing = normalize(m.text || '').slice(0, 100);
    if (!existing) continue;
    if (norm.includes(existing) || existing.includes(norm)) return true;
    const a = new Set(norm.split(' '));
    const b = new Set(existing.split(' '));
    const overlap = Array.from(a).filter((w) => b.has(w)).length;
    if (overlap >= Math.max(3, Math.min(a.size, b.size) / 2)) return true;
  }
  return false;
}

function extractTitle(message: string, lang: string): string {
  const text = message.trim();
  const parts = text.split(/[.!?\n]+/);
  let title = (parts[0] || text).trim();
  title = title
    .replace(/^(σήμερα|χθες|εχθές|today|yesterday|το\s+μωρό|το\s+παιδί|my\s+baby|the\s+baby)\s*[,:-]?\s*/i, '')
    .trim();
  if (title.length > 88) title = `${title.slice(0, 85).trim()}…`;
  if (title.length < 8) title = text.slice(0, 88);
  if (lang === 'el' && title) {
    return title.length > 1 ? title[0].toUpperCase() + title.slice(1) : title.toUpperCase();
  }
  return title ? title[0].toUpperCase() + title.slice(1) : text.slice(0, 88);
}

function inferRef(
  message: string,
  profile: {
    childName?: string;
    dueDate?: string | null;
    children?: { name: string }[];
  },
): string | undefined {
  const msgLower = message.toLowerCase();
  for (const ch of profile.children || []) {
    if (ch.name && msgLower.includes(ch.name.toLowerCase())) return ch.name;
  }
  if (
    profile.dueDate &&
    /(εγκυμοσύν|pregnanc|έμβρυ|κοιλ|bump|ultrasound|υπέρηχ)/i.test(message)
  ) {
    return 'pregnancy';
  }
  const first = profile.children?.[0]?.name;
  if (first) return first;
  return profile.childName || undefined;
}

export function detectMemorySuggestion(
  message: string,
  options: {
    profile?: {
      childName?: string;
      dueDate?: string | null;
      children?: { name: string }[];
    };
    recentMemories?: Pick<AppMemory, 'text'>[];
    lang?: string;
  } = {},
): MemorySuggestion | null {
  const text = (message || '').trim();
  if (text.length < 14 || text.length > 600) return null;
  if (SKIP.test(text)) return null;

  let kind: MemorySuggestionKind | null = null;
  if (MILESTONE.test(text)) kind = 'milestone';
  else if (MOMENT.test(text)) kind = 'moment';
  else if (text.length >= 28 && EMOTIONAL.test(text)) kind = 'moment';

  if (!kind) return null;

  const lang = options.lang || 'el';
  const title = extractTitle(text, lang);
  const recent = options.recentMemories || [];
  if (isDuplicate(title, recent) || isDuplicate(text, recent)) return null;

  const ref = inferRef(text, options.profile || {});
  const description = normalize(text) !== normalize(title) ? text : null;

  return {
    text: title,
    emoji: emojiFor(text),
    description: description ? description.slice(0, 240) : null,
    ref: ref ?? null,
    kind,
    dateIso: new Date().toISOString().slice(0, 10),
  };
}

function wordOverlapScore(a: string, b: string): number {
  const setA = new Set(normalize(a).split(' ').filter(Boolean));
  const setB = new Set(normalize(b).split(' ').filter(Boolean));
  if (!setA.size || !setB.size) return 0;
  const overlap = Array.from(setA).filter((w) => setB.has(w)).length;
  return overlap / Math.max(setA.size, setB.size);
}

/** Fuzzy-match chat text to a milestone checklist label index. */
export function findMatchingMilestoneIndex(text: string, labels: string[]): number | null {
  let bestIdx: number | null = null;
  let bestScore = 0;
  labels.forEach((label, idx) => {
    const score = wordOverlapScore(text, label);
    if (score > bestScore && score >= 0.35) {
      bestScore = score;
      bestIdx = idx;
    }
  });
  return bestIdx;
}

export function mapApiMemorySuggestion(raw: Record<string, unknown> | null | undefined): MemorySuggestion | null {
  if (!raw || typeof raw.text !== 'string' || !raw.text.trim()) return null;
  return {
    text: String(raw.text),
    emoji: typeof raw.emoji === 'string' ? raw.emoji : '💛',
    description: typeof raw.description === 'string' ? raw.description : null,
    ref: typeof raw.ref === 'string' ? raw.ref : null,
    kind: raw.kind === 'milestone' || raw.kind === 'moment' ? raw.kind : undefined,
    dateIso: typeof raw.date_iso === 'string' ? raw.date_iso : new Date().toISOString().slice(0, 10),
  };
}

File: backend/memory_suggestions.py
"""Detect capturable family moments in chat user messages."""

from __future__ import annotations

import re
from datetime import date
from typing import Any, Optional
import unicodedata


def _fold(s: str) -> str:
    """Lowercase + strip accents for fuzzy Greek/ Latin matching."""
    s = (s or "").lower()
    nfkd = unicodedata.normalize("NFD", s)
    return "".join(c for c in nfkd if unicodedata.category(c) != "Mn")

# Questions / help requests — not journal moments
_SKIP = re.compile(
    r"(?i)^(\s*(τι|what|how|why|when|where|ποιος|ποια|ποιο|πώς|γιατί|πότε|που)\b|"
    r".*\?(.*\?)?$|"
    r"(should i|τι να κάνω|help me|βοήθ|μπορείς|can you|tell me|πες μου|explain))",
)

# Strong milestone / first-time signals
_MILESTONE = re.compile(
    r"(?i)(πρώτ|πρωτ|1ο\s|1η\s|first\s+(time|steps?|word|tooth|smile|bath|haircut)|"
    r"πρώτα\s+(βήματα|δοντάκ|χαμόγελ|λέξη|μπιμπίκο|κούρεμα)|"
    r"πρωτα\s+(βηματα|δοντακ|χαμογελ|λεξη|μπικινι|κουρεμα))",
)

# Today/yesterday + achievement verb
_MOMENT = re.compile(
    r"(?i)(σήμερα|σημερα|χθες|εχθές|εχες|today|yesterday).{0,50}"
    r"(έκανε|εκανε|έκαν|εκαν|περπάτη|περπατη|walk|smil|κοιμ|slept|είπε|ειπε|said|έφαγε|εφαγε|ate|έπεσε|επεσε|fell|μπήκε|μπηκε|climb)",
)

# Proud / emotional sharing (softer signal — needs min length)
_EMOTIONAL = re.compile(r"(?i)(🎉|💛|❤️|τόσο\s+όμορφ|so\s+(cute|proud|happy)|μπράβο|proud)")

_EMOJI_MAP = (
    (re.compile(r"(?i)(δοντ|tooth|teeth)"), "🦷"),
    (re.compile(r"(?i)(χαμογ|smil)"), "😊"),
    (re.compile(r"(?i)(βήμα|walk|step|περπα)"), "🚶"),
    (re.compile(r"(?i)(μπάνι|bath)"), "🛁"),
    (re.compile(r"(?i)(κοιμ|sleep)"), "😴"),
    (re.compile(r"(?i)(μίλ|word|speak|said|είπε|babbl)"), "👶"),
)


def _emoji_for(text: str) -> str:
    for pat, em in _EMOJI_MAP:
        if pat.search(text):
            return em
    return "💛"


def _normalize(s: str) -> str:
    return re.sub(r"\s+", " ", (s or "").strip().lower())


def _is_duplicate(text: str, recent: Optional[list[Any]]) -> bool:
    norm = _normalize(text)[:100]
    if not norm:
        return True
    for m in recent or []:
        existing = _normalize(getattr(m, "text", None) or (m.get("text") if isinstance(m, dict) else ""))
        if not existing:
            continue
        if norm in existing or existing in norm:
            return True
        # word overlap
        a = set(norm.split())
        b = set(existing.split())
        if len(a & b) >= max(3, min(len(a), len(b)) // 2):
            return True
    return False


def _extract_title(message: str, lang: str) -> str:
    text = message.strip()
    # First sentence / clause
    parts = re.split(r"[.!?\n]+", text, maxsplit=1)
    title = (parts[0] or text).strip()
    # Trim filler openers
    title = re.sub(
        r"(?i)^(σήμερα|χθες|εχθές|today|yesterday|το\s+μωρό|το\s+παιδί|my\s+baby|the\s+baby)\s*[,:-]?\s*",
        "",
        title,
    ).strip()
    if len(title) > 88:
        title = title[:85].rstrip() + "…"
    if len(title) < 8:
        title = text[:88].rstrip()
    # Capitalize first letter for display
    if title and lang == "el":
        return title[0].upper() + title[1:] if len(title) > 1 else title.upper()
    return title[:1].upper() + title[1:] if title else text[:88]


def _infer_ref(message: str, profile: Any) -> Optional[str]:
    if not profile:
        return None
    children = getattr(profile, "children", None) or []
    msg_lower = message.lower()
    for ch in children:
        name = getattr(ch, "name", None) or (ch.get("name") if isinstance(ch, dict) else None)
        if name and name.lower() in msg_lower:
            return name
    if getattr(profile, "dueDate", None) and re.search(
        r"(?i)(εγκυμοσύν|pregnanc|έμβρυ|κοιλ|bump|ultrasound|υπέρηχ)", message
    ):
        return "pregnancy"
    if children:
        first = children[0]
        return getattr(first, "name", None) or (first.get("name") if isinstance(first, dict) else None)
    child_name = getattr(profile, "childName", None)
    return child_name or None


def detect_memory_suggestion(
    message: str,
    profile: Any = None,
    recent_memories: Optional[list[Any]] = None,
    lang: str = "el",
) -> Optional[dict[str, Any]]:
    """
    Return structured suggestion or None.
    Analyses the USER message (not the assistant reply).
    """
    text = (message or "").strip()
    if len(text) < 14 or len(text) > 600:
        return None
    folded = _fold(text)
    if _SKIP.search(text) or _SKIP.search(folded):
        return None

    kind = None
    if _MILESTONE.search(text) or _MILESTONE.search(folded):
        kind = "milestone"
    elif _MOMENT.search(text) or _MOMENT.search(folded):
        kind = "moment"
    elif len(text) >= 28 and (_EMOTIONAL.search(text) or _EMOTIONAL.search(folded)):
        kind = "moment"

    if not kind:
        return None

    title = _extract_title(text, lang or "el")
    if _is_duplicate(title, recent_memories) or _is_duplicate(text, recent_memories):
        return None

    ref = _infer_ref(text, profile)
    description = text if _normalize(text) != _normalize(title) else None

    return {
        "text": title,
        "emoji": _emoji_for(text),
        "description": (description[:240] if description else None),
        "ref": ref,
        "kind": kind,
        "date_iso": date.today().isoformat(),
    }
